setup.py

```diff
@@ -1,5 +1,9 @@
 from distutils.core import setup
 
+REQUIRES = (
+        'six==1.12.0',
```

Comment:
requiring an exact version of a dependency makes it difficult to use this library in other projects. Using lower bounds is [suggested](https://packaging.python.org/discussions/install-requires-vs-requirements/#install-requires), this way the exact version of the package is up to the projects using this library and reduces the chances of version conflicts.

In this case `string_types` has been in `six` since [1.0.0](https://github.com/benjaminp/six/commit/dfc800b27e5b5b24519087e69766f20f1c16ade0#diff-ec00a44e8e04e1bb5d34c296bdd1e9ebR14).

This would be the first external dependency of the library. I'm not in favor avoiding external dependencies as much as possible.. Specially python2/3 compatibility dependencies are more in question, because Python 2 is almost end of life, and we could just drop Python 2 support (as many other libraries have already done). I preferred the previous implementation that would check the type of the variable with `str`. @slashmili what do you think?

Reply:
still waiting for your prescription for this one

Reply:
My suggestion is not to use `six` at all, and when asserting for type of `sep`, just check the type with `str`. This way:

 * we don't introduce an external dependency
 * we actually keep the behavior closer to Python standard library. Python 3 `str` is Unicode (that's why there is no `unicode` type anymore). Python 2 `str` and `unicode` are different, both inheriting from `basestring`. However Python 2 `isoformat()` does not accept a `unicode`, but only `str`. So we don't need to check with `basestring`/`unicode` types anymore, so we won't need the abstraction of `six`.

```
farzad@farzad-thinkpad ~> python
Python 2.7.15 (default, Oct 15 2018, 15:26:09) 
[GCC 8.2.1 20180801 (Red Hat 8.2.1-2)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from datetime import datetime
>>> datetime.now().isoformat(u'ب')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: isoformat() argument 1 must be char, not unicode
```

Reply:
on the contrary, it does pass unicode actually

here's the local uncommitted code:
```
        assert isinstance(sep, str) and len(sep) == 1, \
            'argument 1 must be a single character: {} -> {}'.format(sep,
                                                                     type(sep))
```

And here's the tests:
```
meysam@MEYSAM:python-jalali$ python3 t/test.py 
.............................................................s...E...s..s
======================================================================
ERROR: test_with_fa_locale (__main__.TestJDateTime)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "t/test.py", line 552, in test_with_fa_locale
    self.set_fa_locale()
  File "t/test.py", line 561, in set_fa_locale
    locale.setlocale(locale.LC_ALL, "fa_IR")
  File "/usr/local/lib/python3.7/locale.py", line 604, in setlocale
    return _setlocale(category, locale)
locale.Error: unsupported locale setting

----------------------------------------------------------------------
Ran 73 tests in 0.222s

FAILED (errors=1, skipped=3)
meysam@MEYSAM:python-jalali$ python t/test.py 
................................................FF..........s....E...s..s
======================================================================
ERROR: test_with_fa_locale (__main__.TestJDateTime)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "t/test.py", line 552, in test_with_fa_locale
    self.set_fa_locale()
  File "t/test.py", line 561, in set_fa_locale
    locale.setlocale(locale.LC_ALL, "fa_IR")
  File "/usr/lib/python2.7/locale.py", line 581, in setlocale
    return _setlocale(category, locale)
Error: unsupported locale setting

======================================================================
FAIL: test_isoformat_custom_timespec (__main__.TestJDateTime)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "t/test.py", line 638, in test_isoformat_custom_timespec
    hours = jdt.isoformat(timespec='hours')
  File "/home/meysam/Codes/python-jalali/jdatetime/__init__.py", line 1219, in isoformat
    type(sep))
AssertionError: argument 1 must be a single character: T -> <type 'unicode'>

======================================================================
FAIL: test_isoformat_default_args (__main__.TestJDateTime)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "t/test.py", line 618, in test_isoformat_default_args
    jiso = jdt.isoformat()
  File "/home/meysam/Codes/python-jalali/jdatetime/__init__.py", line 1219, in isoformat
    type(sep))
AssertionError: argument 1 must be a single character: T -> <type 'unicode'>

----------------------------------------------------------------------
Ran 73 tests in 0.135s

FAILED (failures=2, errors=1, skipped=3)
```

Reply:
> on the contrary, it does pass unicode actually

I meant the `isoformat()` from the standard library. As the shared snippet showed,
it only accepts `str` values, and raises on `unicode` separators. Here is another snippet with more details (checked out from [your branch](https://github.com/meysam81/python-jalali/)), running in a virtualenv with Python 2:

```
(_venv) farzad@farzad-thinkpad ~/t/python-jalali (isoformat/timespec)> python
Python 2.7.15 (default, Oct 15 2018, 15:26:09) 
[GCC 8.2.1 20180801 (Red Hat 8.2.1-2)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from datetime import datetime
>>> sep_str = 'T'
>>> type(sep_str)
<type 'str'>
>>> datetime.now().isoformat(sep_str)  # str is accepted
'2019-07-08T21:48:21.059422'
>>> sep_unicode = u'T'
>>> type(sep_unicode)
<type 'unicode'>
>>> datetime.now().isoformat(sep_unicode)  # unicode is not
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: isoformat() argument 1 must be char, not unicode
>>> 
```
The exception message explicitly says `unicode` is not accepted.

Now, the behavior in the test accepting `unicode` but failing on `str` is actually the opposite of the standard library (desired behavior). We're experiencing this behavior because `jdatetime` [uses UTF-8](https://github.com/slashmili/python-jalali/blob/4e429273e660c8cb3d749bbda229032cd961909c/jdatetime/__init__.py#L1) as source file encoding, so all string literals in the source file are `unicode` not `str`. So the [default value](https://github.com/meysam81/python-jalali/blob/d03f3b920eebb17068f16b859cdda015d42fdde8/jdatetime/__init__.py#L1215) for `sep` parameter (literal "T"), is a `unicode`not a `str`, as clearly demonstrated here:

```
(_venv) farzad@farzad-thinkpad ~/t/python-jalali (isoformat/timespec)> python
Python 2.7.15 (default, Oct 15 2018, 15:26:09) 
[GCC 8.2.1 20180801 (Red Hat 8.2.1-2)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import jdatetime
>>> jdatetime.datetime.isoformat.func_defaults
(u'T', u'auto')
>>> 
```

Reply:
I think I have another way in mind